### ISSUE TITLE
Fixup `impPopList` to eliminate the need to normalize SIMD and HWIntrinsic type handles

### ISSUE THREAD

**Issue by tannergooding** [dotnet/coreclr#21550]
Currently `impPopList` is validating that the struct handle for a given `TYP_SIMD` matches the struct type we have from the stack's type information: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L857

However, in `gtGetStructHandleIfPresent` we do not currently differentiate between SIMD types and HWIntrinsic types for the `GT_IND` and `GT_LCL_FLD` cases:
* https://github.com/dotnet/coreclr/blob/master/src/jit/gentree.cpp#L16446
* https://github.com/dotnet/coreclr/blob/master/src/jit/gentree.cpp#L16460

This means that we have to force normalization for these cases on 9/10 `Vector<T>` types and all of the HWIntrinsic types. It also means that we needed to update `gtGetStructHandleForSIMD` (https://github.com/dotnet/coreclr/pull/21549) to handle the case where we are trying to get a `HWIntrinsic` struct handle for a `TYP_SIMD` that is larger than the size of `Vector<T>` .

It may be beneficial to update this logic so that we are able to get the correct handle and avoid normalization when the types already line up.

**Comment by tannergooding** [comment]
CC. @CarolEidt

**Comment by CarolEidt** [comment]
@tannergooding and I chatted a bit about this offline - to elaborate a bit further: it is beneficial to avoid adding an `OBJ(ADDR)` to SIMD nodes, as they almost never need to be accessed indirectly (the exception is when a parameter needs to be passed as a byref). However, the requirement that struct type nodes map to a valid handle is sprinkled around the JIT, even where it doesn't need to be. And given that ABI passing can depend on the internal structure of the node (see the discussion in https://github.com/dotnet/coreclr/pull/21304), we can't simply eliminate this without ensuring that, at the time the `fgArgTabEntry`s are created, we know which kind of vector we have, not just its size.

**Comment by mikedn** [comment]
>  And given that ABI passing can depend on the internal structure of the node (see the discussion in #21304), we can't simply eliminate this without ensuring that, at the time the fgArgTabEntrys are created, we know which kind of vector we have, not just its size.

Why do we care about the type of the argument node? Shouldn't the type of the method parameter dictate the ABI passing rules?

**Comment by fiigii** [comment]
> However, the requirement that struct type nodes map to a valid handle is sprinkled around the JIT

I think the core problem is `TYP_SIMD* + base_type` cannot be fully mapped to `S.N.vector*` and `Vector64/128/256<T>`.  And we know sometimes `gtGetStructHandleForSIMD` may be called in `Vector64/128/256<T>` senarios that makes things worse. For example, we might get `Vector4`'s hanlde for `Vector128<float>` on AVX2 machines.
https://github.com/dotnet/coreclr/blob/bbfafe73ee054cc024c7db5dea53ec18c25b013c/src/jit/compiler.h#L7632-L7649

**Comment by fiigii** [comment]
If we indeed need to recover struct handle from JIT types, can we add more JIT types for the new hardware SIMD types? e.g., `TYP_HWSIMD16`,  `TYP_HWSIMD32`, etc.

**Comment by CarolEidt** [comment]
> Why do we care about the type of the argument node? Shouldn't the type of the method parameter dictate the ABI passing rules?

Yes, though for structs they must be the same. That said, the JIT currently uses the type of the node, which is not incorrect (it must have the same ABI requirements). My thought is that it would be better to capture this in `impPopList`, as at that point we have both the signature as well as the `ldobj`, which for correctness must (for structs) have the same handle as the signature. However, we don't currently have a way to stash that, as we don't build the `fgArgTabEntry` until morph. I *think* it might be useful to build those during the importer, though we'd still need `fgInitArgInfo` probably to handle the special args. Lots of hand-waving, though, at this point.

**Comment by CarolEidt** [comment]
> I think the core problem is TYP_SIMD* + base_type cannot be fully mapped to S.N.vector* and Vector64/128/256<T>.

Yes, that's the core problem, but I really don't think we need more types - we just need to be sure we capture the correct handle *or* the ABI requirements where it's needed.

**Comment by tannergooding** [comment]
> Shouldn't the type of the method parameter dictate the ABI passing rules?

It might be simpler if we said that `Vector2` maps to `__m64`, `Vector4` maps to `__m128`, and `Vector<T>` maps to one of `__m128` or `__m256` (depending on its size). That wouldn't work for `Vector3` (TYP_SIMD12), however.

However, I think there could also be some other problems with treating the existing `S.N.Vector` types as interchangeable with the `HWIntrinsic` types, namely around the layout/packing of structs. For the HWIntrinsic types, we made sure to update the VM so that these types would have the appropriate packing. We did not do the same for the S.N.Vector types and changing it at this point (several releases after shipping) would be breaking.

**Comment by mikedn** [comment]
> S.N.Vector types as interchangeable with the HWIntrinsic types, namely around the layout/packing of structs.

It's unlikely that layout will cause problems in the JIT, that's more or less VM's concern.

In general, I don't see any serious problems in treating existing vector types the same as the new vector types. So far, the only obvious problem we have is caused by calling convention differences but there must be a better way to deal with that.

> I think it might be useful to build those during the importer, though we'd still need fgInitArgInfo probably to handle the special args.

Doing it in the importer might harm throughput because some work may go to waste due to inlining. I think it's fine to do this during morph (or maybe in lowering, but that's another story), the only problem is using the types from the call signature and not the types from the argument nodes.

> If we indeed need to recover struct handle from JIT types, can we add more JIT types for the new hardware SIMD types? e.g., TYP_HWSIMD16, TYP_HWSIMD32, etc.

Ugh, that will just make the code more complicated. And long term it will probably make it more difficult to replace the current SIMD codegen with HWIntrinsic codegen to get rid of duplication. Also, you need to be a bit careful with adding more types, here and there the JIT uses some tables indexed by type and adding new types will make those tables larger, large enough to show up in mem stats.
